File: HW/【队列】2023B-不开心的小朋友.java
/*
 * @hw id=2023B lang=java
 *
 * 不开心的小朋友
 */

// @hw code=start
import java.util.*;
public class Main {
    public static void main(String[] args) {
        Scanner in = new Scanner(System.in);
        int n = in.nextInt();
        String s = in.nextLine();
        String[] arr = s.split(" ");
        Queue<Integer> wq = new LinkedList<>();
        Map<String, List<Boolean>> state = new HashMap<>();
        int playNum = 0;
        int angryNum = 0;
        for(String c : arr) {
            if(!state.containsKey(c)) {
                state.put(c, new ArrayList<>(false, false, false));
            }
            if(!state.get(c).get(0)) {
                state.get(c).set(0, true);
                if(playNum>0) {
                    state.get(c).set(1, true);
                    playNum++;
                } else {
                    wq.add(c);
                }
            } else {
                state.get(c).set(2, true);
                if(state.get(c).get(1)) {
                    playNum--;
                    while(!wq.isEmpty() && playNum < n) {
                        String nextc = wq.poll();
                        if(!state.get(nextc).get(2)) {
                            state.get(nextc).set(1, true);
                            playNum++;
                        }
                    }
                } else {
                    angryNum++;
                }
            }
        }
    }
    System.out.println(angryNum);
}
// @hw code=end

File: HW/【队列】2023B-不开心的小朋友.py
#
# @hw id=2023B lang=python3
#
# 不开心的小朋友
#

# @hw code=start
n = int(input())
cs = input().split()
wq = list()
state = {}
playNum = 0
angryNum = 0 

for c in cs:
    if c not in state:
        # 0 是否出现 1 是否在玩 2 是否离开
        state[c] = [False, False, False]
    # 第一次出现
    if not state[c][0]:
        # 标记出现
        state[c][0] = True
        if playNum < n:
            # 有空位
            playNum += 1
            state[c][1] = True
        else:
            wq.append(c)
    else:
        # 第二次出现
        state[c][2] = True
        if state[c][1]:
            playNum -= 1
            while playNum < n and len(wq) > 0:
                nc = wq.pop(0)
                if not state[nc][2]:
                    state[nc][1] = True
                    playNum += 1
        else:
            angryNum += 1
print(angryNum)

# @hw code=end